### DISCUSSION TITLE
Help needed with creating 301 redirects in Nuxt 3

### DISCUSSION THREAD

**Discussion by toddpadwick** [nuxt/nuxt#16088] [Questions] [answered]
Hey everyone. In Nuxt 2, I would run an Axios query using NuxtServerInit in dev mode, to gather redirects from an API, store them as a redirects.json file, and then imported them into a server middleware function like so:
```
import redirects from './assets/redirects.json';

module.exports = function (req, res, next) {
  const url = req.url;
  const redirect = redirects.find(obj => new RegExp(`^${obj.from}`).test(url));
  if (redirect) { // if redirect exists
    console.info('Redirect found.');
    const newLocation = redirect.to // redirect the newlocation
    res.writeHead(parseInt(redirect.statusCode), {
      Location: newLocation
    })
    res.end()
  } else { 
    next()
  }
}

```

There is a lot of documentation and guidance on doing this in Nuxt 2. But very little information on Nuxt 3. I have heard some light discussion about this being a core feature using Nuxt 3 route rules,but I cannot find any documentation on doing so for redirects. Can anyone help me?

**Comment by toddpadwick** [comment, marked as answer]
UPDATE - PROGRESS MADE BUT FURTHER HELP STILL REQUIRED: I have since managed to figure out a possible solution, but I've no idea if this is the right way or not. I also don't know if this has any effect on performance in production or whether it will work okay in both static and universal modes. Can anyone check if this is a viable method?

*Nuxt.config.ts*
```
import { resolve } from 'path';

import sanityClient from '@sanity/client' // Using the Sanity CMS API fetch query

export const getRedirects = async () => {
	let redirectsObject = {}; // store empty redirects object as default

	const redirectQuery = `*[_type == "redirect"]{
		'from':fromPath,
		'redirect':{'to':toPath, 'statusCode':statusCode}
	}`; // setting up the Sanity CMS query using GROQ

	const client = sanityClient({
		projectId: process.env.SANITY_PROJECT_ID,
		apiVersion: process.env.SANITY_API_VERSION,
		useCdn:process.env.SANITY_CDN,
		dataset:process.env.SANITY_DATASET
	}); // storing client fetch credentials

	const data = await client.fetch(redirectQuery); // run the query

	if (data && data.length) {
		data.forEach(item => { // Run a loop to create an object from the array as based on the docs, it seems the routeRules only accepts an object, which seems strange to me as surely most redirects would be initially created as an array
			redirectsObject[item.from] = {
				redirect:item.redirect
			}
		})
		console.info('Redirects are generated');
	} else {
		console.info('No redirects found in Sanity');
	}
	return redirectsObject;
}

export default defineNuxtConfig({
	routeRules: getRedirects() 
})


```

Thanks in advance!

**Comment by bertyn99** [comment]
I dont know the best way to do redirection i did a lot of research but nothing very clear for now i did  like that with a route middleware:
```js 
export default defineNuxtRouteMiddleware((to, from) => {
//array of object with source and destination of redirection
  const listRedirects = [
    { from: "/blog/basic-of-javascript", to: "/blog/what-is-javascript" },
  ];

//filter the redirect object to correspong  to the  path of the current request
  let redirectPath = listRedirects.filter((v) => v.from == to.path);

//if it is then redirect seems to work with google crawler  checker with search console
  if (redirectPath.length !== 0 && redirectPath[0].to) {
    return navigateTo(redirectPath[0].to, { redirectCode: 301 });
  }
});

 ```

**Reply by andreavettore** [reply]
Hi @bertyn99 
is it possible to handle redirects on server side with Nuxt3, using /server/middleware folder?

**Reply by toddpadwick** [reply]
From what I gather, with Nuxt3, the recommended approach is to use RouteRules rather than route middleware. Route Middleware was the approach I used to use for Nuxt 2.

**Reply by andreavettore** [reply]
Hi @toddpadwick 
yeah, I think RouteRules are very cool but I have to handle a very big amount of redirects and I don't want to affect the build time on Netlify. The redirects are processed in there.

**Reply by bertyn99** [reply]
my snippet work in server/middleware and work with server side rendering

**Reply by bertyn99** [reply]
for routes rules i dont understand at all how to use it

**Reply by andreavettore** [reply]
> my snippet work in server/middleware and work with server side rendering

are you sure? I tried to put your snippet on server/middleware/handle-redirects.js and I'm getting an error:
defineNuxtRouteMiddleware is not defined

**Reply by andreavettore** [reply]
> for routes rules i dont understand at all how to use it

in your nuxt.config.ts

https://nitro.unjs.io/config/#routerules

**Reply by bertyn99** [reply]
my bad with fatigue i  made a mistake  my snippet needed a little tweek

**Reply by bertyn99** [reply]
Just the normal middleware  and  server/middleware are a little different with my understanding 
Route middleware run within the Vue part of your Nuxt app. Despite the similar name, they are completely different from server middleware, which are run in the Nitro server part of your app.
 so if i go to for example with nuxt content  with 2 middleware that log the request 
`localhost:3000/blog/article`

- Client middleware  1 log
`New request /blog/article`

- Server middleware  2 log
`New request /blog/article` (client navigator)
`New request /api/_content/query?_params={query:/blog/article}` (api nuxt content)

route Client middleware work with SSR when i tested it 
So are you sure you need Server middleware ?
